qiskit/transpiler/passes/mapping/check_cnot_direction.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""
+This pass checks if the CNOTs (or any other 2Q) in the DAG follow the right
+direction with respect to thecoupling map.
+"""
+
+from qiskit.transpiler._basepasses import AnalysisPass
+from qiskit.mapper import Layout
+from qiskit.extensions.standard.swap import SwapGate
+
+
+class CheckCnotDirection(AnalysisPass):
+    """
+    checks if the CNOTs (or any other 2Q) in the DAG follow the right
+    direction with respect to thecoupling map
+    """
+
+    def __init__(self, coupling_map, initial_layout=None):
+        """
+        Checks if the 2Q gates in DAGCircuit are in the allowed direction with
+        respect to `coupling_map`.
+        Args:
+            coupling_map (CouplingMap): Directed graph represented a coupling map.
+            initial_layout (Layout): The initial layout of the DAG to analyze.
+        """
+        super().__init__()
+        self.layout = initial_layout
+        self.coupling_map = coupling_map
+
+    def run(self, dag):
+        """
+        If `dag` is mapped and the direction is correct the property
+        `is_direction_mapped` is set to True (or to False otherwise).
+
+        Args:
+            dag (DAGCircuit): DAG to map.
+        """
+        if self.layout is None:
+            if self.property_set["layout"]:
+                self.layout = self.property_set["layout"]
+            else:
+                self.layout = Layout.generate_trivial_layout(*dag.qregs.values())
+
+        self.property_set['is_direction_mapped'] = True
+
+        for gate in dag.get_2q_nodes():
+            physical_q0 = self.layout[gate['qargs'][0]]
+            physical_q1 = self.layout[gate['qargs'][1]]
+
+            if (physical_q0, physical_q1) not in self.coupling_map.get_edges():
+                self.property_set['is_direction_mapped'] = False
+                return
+
+            if isinstance(gate['op'], SwapGate):
```

Comment:
A backwards-mapped`SwapGate` will hit the early `return` above, so I think this code will never be run.

Reply:
On `SwapGate`s, how should they be handled here? Should any `swap` on adjacent qubits satisfy `is_direction_mapped`, and then when we (maybe) unroll to `cx`s we make sure to do so only with `cx`s from the coupling map? Or do we require `swap`s already be unrolled before we `CheckDirectionMap`?

Reply:
They way I have it, in case of `swap q1, q2`, the coupling map should have `q1 <-> q2`.

Reply:
This pass should only check CNOTs. Not any other 2Q gate. That will not make sense, given that the coupling_map is specifying CNOT directions on the device.

Swap and CZ are not directional gates anyway. They are symmetric.

Reply:
okey.. now ignores everything that is not CxBase. Fixed in 2a806c91